libraries/bot-azure/src/main/java/com/microsoft/bot/azure/blobs/BlobsStorage.java

```diff
@@ -3,5 +3,187 @@
 
 package com.microsoft.bot.azure.blobs;
 
-public class BlobsStorage {
+import com.azure.core.exception.HttpResponseException;
+import com.azure.core.http.rest.Response;
+import com.azure.core.util.Context;
+import com.azure.storage.blob.BlobClient;
+import com.azure.storage.blob.BlobContainerClient;
+import com.azure.storage.blob.BlobContainerClientBuilder;
+import com.azure.storage.blob.models.BlobErrorCode;
+import com.azure.storage.blob.models.BlobRequestConditions;
+import com.microsoft.bot.builder.Storage;
+import com.microsoft.bot.builder.StoreItem;
+import com.microsoft.bot.restclient.serializer.JacksonAdapter;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.HttpStatus;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+public class BlobsStorage implements Storage {
+    private final JacksonAdapter jacksonAdapter;
+    private final BlobContainerClient containerClient;
+
+    public BlobsStorage(String dataConnectionString, String containerName) {
+        if (StringUtils.isBlank(dataConnectionString)) {
+            throw new IllegalArgumentException("dataConnectionString is required.");
+        }
+
+        if (StringUtils.isBlank(containerName)) {
+            throw new IllegalArgumentException("containerName is required.");
+        }
+
+        jacksonAdapter = new JacksonAdapter();
+        containerClient = new BlobContainerClientBuilder()
+                            .connectionString(dataConnectionString)
+                            .containerName(containerName)
+                            .buildClient();
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(String[] keys) {
+        if (keys == null) {
+            throw new IllegalArgumentException("The 'keys' parameter is required.");
+        }
+
+        for (String key : keys) {
+            String blobName = GetBlobName(key);
+            BlobClient blobClient = containerClient.getBlobClient(blobName);
+            blobClient.delete();
+        }
+
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public CompletableFuture<Map<String, Object>> read(String[] keys) {
+        if (keys == null) {
+            throw new IllegalArgumentException("The 'keys' parameter is required.");
+        }
+
+        initialize();
+
+        Map<String, Object> items = new HashMap<>();
+
+        for (String key : keys) {
+            try {
+                String blobName = GetBlobName(key);
+                BlobClient blobClient = containerClient.getBlobClient(blobName);
+                items.put(key, innerReadBlob(blobClient).get());
```

Comment:
Is the `get` ok?

Reply:
Mmm.... no, it should be join 👍

Reply:
Nope, the join will stop the main thread, we need to get the result of the `innerReadBlob` method to assign it inside the `items` list